lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -1657,6 +1666,80 @@ selectionToUnsignedTx wdrl sel s =
         WithdrawalExternal acct path c ->
             [(acct, c, path)]
 
+data CoinSelection = CoinSelection
+    { inputs :: [(TxIn, TxOut, NonEmpty DerivationIndex)]
+    , outputs :: [TxOut]
+    , change :: [TxChange (NonEmpty DerivationIndex)]
+    , collateral :: [(TxIn, TxOut, NonEmpty DerivationIndex)]
+    , withdrawals :: [(RewardAccount, Coin, NonEmpty DerivationIndex)]
+    , delegationAction :: Maybe (DelegationAction, NonEmpty DerivationIndex)
+    , deposit :: Maybe Coin
+    , refund :: Maybe Coin
+    }
+    deriving (Generic)
+
+
+buildCoinSelectionForTransaction
+    :: forall s k n era
+     . ( Cardano.IsCardanoEra era
+       , IsOurs s Address
+       , s ~ SeqState n k
+       )
+    => Wallet s
+    -> [TxOut] -- ^ payment outputs to exclude from change outputs
+    -> Coin
+    -> Maybe DelegationAction
+    -> Cardano.Tx era
+    -> CoinSelection
+buildCoinSelectionForTransaction
+    wallet paymentOutputs depositRefund delegationAction cardanoTx =
+    CoinSelection
+    { inputs = resolveInput . fromCardanoTxIn . fst =<< txIns
+    , outputs = paymentOutputs
+    , change = do
+        out <-
+            -- NOTE: We assume that the change outputs are always
+            -- at the end of the list. This is true for the current
+            -- 'balanceTransaction' implementation, but may not
+            -- be true for other implementations.
+            drop (length  paymentOutputs) $ fromCardanoTxOut <$> txOuts
+        let address = out ^. #address
+        derivationPath <-
+            maybeToList $ fst $ isOurs address (getState wallet)
+        pure TxChange
+            { address
+            , amount = out ^. #tokens . #coin
+            , assets = out ^. #tokens . #tokens
+            , derivationPath
+            }
+    , collateral = resolveInput . fromCardanoTxIn =<<
+        case txInsCollateral of
+            Cardano.TxInsCollateralNone -> []
+            Cardano.TxInsCollateral _supported is -> is
+    , withdrawals = fromCardanoWdrls txWithdrawals <&>
+        \(acct, coin) -> (acct, coin, rewardAcctPath)
```

Comment:
Assuming only one rewardAcctPath seem to be the cause of this integration test failure:
```
  integration/src/Test/Integration/Scenario/API/Shelley/CoinSelections.hs:293:18:
  1) API Specifications, SHELLEY_COIN_SELECTION, WALLETS_COIN_SELECTION_06b - can redeem rewards from other
       predicate failed on: [ApiCoinSelectionWithdrawal {stakeAddress = (ApiT (RewardAccount "\137\252a\210\GS\223\203\212\212\&6R\191\ENQ\196\f4o\167\148\135\DC4#\182PR\215aL"),Proxy), derivationPath = ApiT (DerivationIndex 2147485500) :| [ApiT (DerivationIndex 2147485463),ApiT (DerivationIndex 2147483648),ApiT (DerivationIndex 2),ApiT (DerivationIndex 0)], amount = Quantity 1000000000000}]
       While verifying value:
         ( Status
```

Old code

https://github.com/input-output-hk/cardano-wallet/blob/14bc6b0c2531ce4a2f5e9fcc8174aaa10ed3ec2c/lib/wallet/src/Cardano/Wallet.hs#L1651-L1658